kong/clustering/services/sync/rpc.lua

```diff
@@ -173,7 +176,101 @@ local function is_rpc_ready()
 end
 
 
-local function do_sync()
+local function lmdb_purge()
+  ngx_log(ngx_INFO, "purging the LMDB")
+
+  local t = assert(txn.begin(512))
+  assert(t:db_drop(false))
+  assert(t:set(DECLARATIVE_HASH_KEY, DECLARATIVE_EMPTY_CONFIG_HASH))
+  assert(t:commit()) -- not much we can do if this fails
+
+  -- we are at a unready state
+  -- consider the config empty
+  assert(kong.core_cache:purge())
+  assert(kong.cache:purge())
+end
+
+
+local function reset_dp()
+  ngx_log(ngx_INFO, "resetting the data plane")
+
+  lmdb_purge()
+  assert(build_router("init"))
+end
+
+
+local do_sync
+
+
+local function error_handler(err, is_full_sync)
+  ngx_log(ngx_ERR, err)
+
+  if is_full_sync then
+    -- in case of full sync, reset the DP to initial state and retry immediately
+    reset_dp()
+    return do_sync() -- retry without releasing the mutex as the DP has nothing to do except syncing
+  end
+
+  return nil, err
+end
+
+
+local function lmdb_upsert(db, t, delta, opts)
+  local delta_type = delta.type
+  local delta_entity = delta.entity
+
+  -- upsert the entity
+  -- does the entity already exists?
+  local old_entity, err = db[delta_type]:select(delta_entity)
+
+  if err then
+    return nil, err
+  end
+
+  if old_entity then
+    local res, err = delete_entity_for_txn(t, delta_type, old_entity, opts)
+    if not res then
+      return nil, err
+    end
+  end
+
+  local res, err = insert_entity_for_txn(t, delta_type, delta_entity, opts)
+  if not res then
+    return nil, err
+  end
+
+  return { delta_type, old_entity and "update" or "create", delta_entity, old_entity, }
```

Comment:
Since we are using `"update"`, should we rename this function to `lmdb_update`?

Reply:
It *IS* upsert. It will do insert if the entity does not exist

Reply:
I know it, but it is not consistent with the event "update"？It will confuse the other reviewers.
For now I suggest to use `lmdb_update`, after this PR is merged we could rename it back again.